husky-misc/code-challenge
# Russian Doll Cache Soluction

## What is the purpose of this Pull Request?

This PR solves the challenge of implementing a Russian Doll cache strategy for bank statements.

## What was done to achieve this?

Only that code! It seems easy but it took a time to get to the solution. 

```ruby

    def cache_block(record:, query:, expiration: 12.hours)
      key = "#{record.cache_key_with_version}_#{query}"
      cached_block = Rails.cache.read(key)
      return cached_block unless cached_block.nil?
      cached_block = yield.to_a
      Rails.cache.write(key, cached_block, expires_in: expiration)
      cached_block
    end

```
## How to test that this really works?

This soluction focus on the goal. First lets will start the project (locally).

step | description
-- | --
1 | Clone this repository.
2 | Make sure the installation of PostgreSQL (9.xx) in your machine.
3 | Make sure the installation of [Redis](https://www.hostinger.com.br/tutoriais/install-redis-ubuntu/) in your machine.
4 | In .env.development and .env.test (files) setup the enviroments variables(USER_DB and  PASSWORD_DB) with your the database user and database password
5 | Create the databases (test and development) and execute migrations ( rails db:create && rails db:migrate and rails db:create RAILS_ENV=test && rails db:migrate RAILS_ENV=test)
6 | Seed database with rails db:seed (its can be take time, we need many transactions to see difference between load data from database or cache memory )
8 | Execute the server (rails s)
9 | Get id of created Bank Account  in this endpoint **GET /api/v1/bank_accounts/**
10 | With id of Bank Account we can generate the Bank Statement **GET /api/v1/bank_accounts/:bank_account_id/bank_statements?days=30** . If we repeate the request the database query for transactions does not will ocurr again. The response time should be faster (than the first time of request, we should see in log of server).
11 | The extract is paginated, even so the result of each page will be cached. (**GET /api/v1/bank_accounts/:bank_account_id/bank_statements?days=30&page%5Bnumber%5D=4683&page%5Bsize%5D=20_**)
12 | A new transaction was added? (**POST /api/v1/bank_accounts/:bank_account_id/transactions**). When we call api again (the request of step 10). The cache will be updated (the old cache not will be used anymore and will expire).
13 |  If all cached data goes away, there is no problem when request for bank statement was made again, the cache will be load. We can improve this caching of code block with a job(sidekiq) or rake easyli.
14 | The unit test for the above code is in spec / lib / caching / active_record_collection_spec.rb.

### Goal
Implement a Russian Doll (or any better strategy) cache for a bank statement receipt

### Expected result

1 - We'd like to see a POC, including some Unit tests that prove the chosen cache strategy.

2 - We'd like to see some stress test consuming your API endpoint that serves the bank statement JSON

3 - Memory and cloud storage are currently cheap. We expect you to use a production-ready infrastructure like Memcached or Redis.

4 - An isolated module that we can use as a DSL to evict or include new cache keys reloading current data

5 - Is there a recovery strategy? And if all cached data goes away, how do we query and cache everything back?

6 - You'll explain to the team your approach and defend the pros and cons of your strategy. There is no silver bullet, and we expect that you will be able to explain to us what are the stronger and the weaker points of the chosen strategy.
It should be a piece of code that you studied a little to figure out and are proud of enough to explain to other people.
